Add AWS collector and recipe

giovannt0 opened the pull request:
Add a new `aws_forensics` recipe and a new `AWSCollector` collector, that offers similar functionalities to the `gcp_forensics` recipe and its corresponding `GoogleCloudCollector` collector.

Marking as draft until libcloudforensics new pip release is out.

Signed-off-by: Theo Giovanna <gtheo@google.com>

tomchop commented on dftimewolf/lib/collectors/aws.py:
```diff
@@ -0,0 +1,252 @@
+    """
+
+    if not (remote_profile_name and remote_zone):
+      self.state.AddError('You must specify "remote_profile_name" and "zone" '
+                          'parameters', critical=True)
+
```
You need to return right after adding the error (pending https://github.com/log2timeline/dftimewolf/issues/80)

tomchop commented on dftimewolf/lib/collectors/aws.py:
```diff
@@ -0,0 +1,252 @@
+    AWS recommends using device names that are within /dev/sd[f-p][1-6].
+
+    Returns:
+      str: A device name.
```
Or an empty string if the device name could not be obtained?

tomchop commented on dftimewolf/lib/collectors/aws.py:
```diff
@@ -0,0 +1,252 @@
+    self.analysis_profile_name = analysis_profile_name or remote_profile_name
+
+    # See https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/device_naming.html
+    self.device_names = [chr(i) for i in range(ord('f'), ord('p') + 1)]
```
You can also do

```python
self.device_names = list('fghijklmnop')
```

I'd also rename this attribute to `device_suffixes`

tomchop commented on dftimewolf/lib/collectors/aws.py:
```diff
@@ -0,0 +1,252 @@
+        self.state.AddError(
+            'Volume "{0:s}" was not found in AWS account {1:s}'.format(
+                volume_id, self.remote_profile_name),
+            critical=True)
```
If the error is critical, then you will want to return an empty list, and check for the return value of this in the caller (I know, I should really fix #80)

tomchop commented on dftimewolf/lib/collectors/aws.py:
```diff
@@ -0,0 +1,252 @@
+
+    Args:
+      instance_id (str): Instance ID of the instance to get the volumes from.
+      all_volumes (bool): If set, get all volumes attached to the instance. If
```
```suggestion
      all_volumes (bool): If True, get all volumes attached to the instance. If
```

tomchop commented on dftimewolf/lib/collectors/aws.py:
```diff
@@ -0,0 +1,252 @@
+    return [remote_instance.GetBootVolume()]
+
+  def _FindVolumesToCopy(self):
+    """Determines which volumes to copy depending on object attributes.
```
```suggestion
    """Determines which volumes to copy depending on the collector's attributes.
```

giovannt0 commented:
Placing this on hold until https://github.com/google/cloud-forensics-utils/issues/159 is resolved, since it will likely change the collector.

giovannt0 commented:
@tomchop PTAL, I updated the code to work with the new libcloudforensics release and addressed your comments. Note that the pinned dependency for libcloudforensics should be updated first (to 20200630 or more recent).

tomchop commented on data/recipes/aws_forensics.json:
```diff
@@ -0,0 +1,33 @@
+{
+  "name": "aws_forensics",
+  "short_description": "Copies volume from an AWS account to an analysis VM.",
```
```suggestion
  "short_description": "Copies a volume from an AWS account to an analysis VM.",
```

tomchop commented on dftimewolf/lib/collectors/aws.py:
```diff
@@ -0,0 +1,255 @@
+        source account in which the volume(s) exist(s).
+    incident_id (str): Incident identifier used to name the analysis VM.
+    remote_instance_id (str): Instance ID that needs forensicating.
+    volume_ids (list[str]): List of volume ids to copy.
```
```suggestion
    volume_ids (list[str]): List of volume IDs to copy.
```

tomchop commented on dftimewolf/lib/collectors/aws.py:
```diff
@@ -0,0 +1,255 @@
+      cpu_cores (int): Optional. The number of CPU cores to use for the
+          analysis VM. Default is 16.
+      ami (str): Optional. The Amazon Machine Image ID to use to create the
+          analysis VM. Default is Ubuntu 18.04 TLS.
```
It seems like the default parameter here is None. So maybe rephrase this to say `if not specified, will default to selecting Ubuntu 18.04 LTS`.

tomchop commented on dftimewolf/lib/collectors/aws.py:
```diff
@@ -0,0 +1,255 @@
+          new_volume, self._FindNextAvailableDeviceName())
+      print('Volume {0:s} successfully copied to {1:s}'.format(
+          volume.volume_id, new_volume.volume_id))
+      self.state.output.append((self.analysis_vm.name, new_volume))
```
Since #279 we're not using input / output anymore. Please replace this with a a `ForensicsVM` attribute container like in [gcloud.py](https://github.com/log2timeline/dftimewolf/blob/master/dftimewolf/lib/collectors/gcloud.py#L70)

tomchop commented on dftimewolf/lib/collectors/aws.py:
```diff
@@ -0,0 +1,255 @@
+    self.analysis_profile_name = analysis_profile_name or remote_profile_name
+
+    # See https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/device_naming.html
+    self.device_suffixes = list('fghijklmnop')
```
Shouldn't this go in `__init__`?

tomchop commented on dftimewolf/lib/collectors/aws.py:
```diff
@@ -0,0 +1,255 @@
+    Returns:
+      list[AWSVolume]: A list of the volumes to copy.
+    """
+    if not (self.remote_instance_id or self.volume_ids):
+      self.state.AddError(
+          'You need to specify at least an instance name or volume ids to copy',
+          critical=True)
+      return []
```
This should go in `SetUp`. We want to be able to fail early if these parameters are not provided. (Should also be changed in `gcloud.py`, #297 )

giovannt0 replied:
Thanks, opened #298 for `gcloud.py`
